### PR TITLE
Bayesian t-test error handling

### PR DESCRIPTION
**Pull request by koenderks**
Implemented error handling for all Bayesian t-tests

### CODE REVIEW

**Comment by boutinb** [comment]
@TimKDJ Did you review this Pull Request?

File: JASP-Engine/JASP/R/commonsummarystatsttestbayesian.R [hunk 1]
```diff
@@ -1374,6 +1374,8 @@
 		nullInterval <- c(-Inf, 0)
 		stretch <- 1.32
 	}
+    
+    BF <- as.numeric(BF)
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/commonsummarystatsttestbayesian.R]: Can you add a comment explaining this? We will need to provide a proper fix in the long term.

File: JASP-Engine/JASP/R/ttestbayesianindependentsamples.R [hunk 2]
```diff
@@ -483,43 +498,29 @@ TTestBayesianIndependentSamples <- function(dataset=NULL, options, perform="run"
...
-							}
-							
+						if (!is.null(errorMessage)) {
+						    
+						    plot[["error"]] <- list(error="badData", errorMessage=errorMessage)
+						    
 						} else {
 							
-							plot[["error"]] <- list(error="badData", errorMessage=plottingError[statusInd])
+						    p <- try(silent= FALSE, expr= {
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/ttestbayesianindependentsamples.R]: We will still want to catch this try. Otherwise a panel might be empty if an unexpected error occurs.

File: JASP-Engine/JASP/R/ttestbayesianindependentsamples.R [hunk 3]
```diff
@@ -1022,6 +998,16 @@ TTestBayesianIndependentSamples <- function(dataset=NULL, options, perform="run"
...
+				                         all.target = variable,
+				                         observations.amount = '< 2')
+				    
+				    if (!identical(errors, FALSE)) {
+				        errorMessage <- errors$message
+				    } else {
+				        errorMessage <- NULL
+				    }
 					
 					result <- try (silent=FALSE, expr= {
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/ttestbayesianindependentsamples.R]: If we find errors we do not want to run this at all.
Additionally, we need the catch structure for exceptions.

File: JASP-Engine/JASP/R/ttestbayesianindependentsamples.R [hunk 4]
```diff
@@ -376,9 +383,16 @@ TTestBayesianIndependentSamples <- function(dataset=NULL, options, perform="run"
 			z <- 1
 			descriptInd <- 1
 			
-			
 			for (variable in options[["variables"]]) {
 				
+			    errors <- .hasErrors(dataset, perform, message = 'short', type = c('infinity','observations','variance'),
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/ttestbayesianindependentsamples.R]: Need to specify the grouping variable

File: JASP-Engine/JASP/R/ttestbayesianindependentsamples.R [hunk 5]
```diff
@@ -1022,6 +998,16 @@ TTestBayesianIndependentSamples <- function(dataset=NULL, options, perform="run"
 					plottingError[rowNo] <- state$plottingError[index]
 					
 				} else {
+
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/ttestbayesianindependentsamples.R]: Add the grouping variable

File: JASP-Engine/JASP/R/ttestbayesianonesample.R [hunk 6]
```diff
@@ -634,33 +634,19 @@ TTestBayesianOneSample <- function(dataset=NULL, options, perform="run", callbac
           list(Variable=variable, BF=BF, error=error)
         })
         
+        errors <- .hasErrors(dataset, perform, message = "short", type = c('observations', 'variance', 'infinity'),
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/ttestbayesianonesample.R]: Shouldn't this go above calculating the model?

File: JASP-Engine/JASP/R/ttestbayesianonesample.R [hunk 7]
```diff
@@ -748,23 +741,24 @@ TTestBayesianOneSample <- function(dataset=NULL, options, perform="run", callbac
           
           plot <- descriptivesPlots[[descriptInd]]
           
+          if (!is.null(errorMessage)) {
+              
+              plot[["error"]] <- list(error="badData", errorMessage=errorMessage)
+              
+          } else {
+          
           p <- try(silent= FALSE, expr= {
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/ttestbayesianonesample.R]: Add a catch for the try, e.g.

```
if (is.null(errorMessage) {
  try plot
  if (isTryError(plot) {
    errorMessage <- .extractErrorMessage(plot)
  }
}
if (! is.null(errorMessage)) {
  log it
}
```

File: JASP-Engine/JASP/R/ttestbayesianonesample.R [hunk 8]
```diff
@@ -825,44 +819,32 @@ TTestBayesianOneSample <- function(dataset=NULL, options, perform="run", callbac
...
-              } else if (errorMessage == "'from' cannot be NA, NaN or infinite") {
-                
-                errorMessage <- "Plotting is not possible: The Bayes factor is too small"
+              if (!is.null(errorMessage)) {
+                  
+                  plot[["error"]] <- list(error="badData", errorMessage=errorMessage)
+                  
+              } else {
+                  
+                  p <- try(silent= FALSE, expr= {
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/ttestbayesianonesample.R]: Same as above.

File: JASP-Engine/JASP/R/ttestbayesianonesample.R [hunk 9]
```diff
@@ -1004,7 +974,13 @@ TTestBayesianOneSample <- function(dataset=NULL, options, perform="run", callbac
...
-          } else if (status[i] != "error" && status[i] != "sequentialNotPossible") {
+          }
+          
+          if (!is.null(errorMessage)) {
+              
+              plot[["error"]] <- list(error="badData", errorMessage=errorMessage)
+              
+          } else {
             
             p <- try(silent= FALSE, expr= {
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/ttestbayesianonesample.R]: And here. We need to still be able to handle exceptions.

File: JASP-Engine/JASP/R/ttestbayesianonesample.R [hunk 10]
```diff
@@ -2688,7 +2646,8 @@ TTestBayesianOneSample <- function(dataset=NULL, options, perform="run", callbac
   # display BF10 value
   if (idData < length(BF10)) {
     
-    BF10e <- BF10post
+      BF10post <- as.numeric(BF10post)
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/ttestbayesianonesample.R]: Same issue that needs fixing I guess? Please add a little fixme note.

File: JASP-Engine/JASP/R/ttestbayesianpairedsamples.R [hunk 11]
```diff
@@ -543,23 +543,29 @@ TTestBayesianPairedSamples <- function(dataset=NULL, options, perform="run", cal
...
-							
-						} else if (errorMessage == "data are essentially constant") {
-							
-							errorMessage <- paste("Bayes factor is undefined - the sample contains all the same value (zero variance)")
-							
-						} else if (errorMessage == "Insufficient sample size for t analysis." || errorMessage == "not enough observations") {
-							
-							errorMessage <- "Bayes factor is undefined - one or both of the variables has too few observations (possibly only after missing values are excluded)"	
-						}
+					errors <- .hasErrors(dataset, perform, message = 'short', type = c('infinity','observations','variance'),
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/ttestbayesianpairedsamples.R]: Above the analysis. If we know we cannot calculate the model, we shouldn't try to.
Also probably the trycatch thing again.

File: JASP-Engine/JASP/R/ttestbayesianpairedsamples.R [hunk 12]
```diff
@@ -742,6 +757,12 @@ TTestBayesianPairedSamples <- function(dataset=NULL, options, perform="run", cal
...
 					
+					if (!is.null(errorMessage)) {
+					    
+					    plot[["error"]] <- list(error="badData", errorMessage=errorMessage)
+					    
+					} else {
+					
 					p <- try(silent= FALSE, expr= {
 
 							p <- .plot2GroupMeansBayesIndTtest(v1 = c1, v2 = c2, nameV1 = pair[[1]], nameV2 = pair[[2]], descriptivesPlotsCredibleInterval=options$descriptivesPlotsCredibleInterval)
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/ttestbayesianpairedsamples.R]: trycatch :)

File: JASP-Engine/JASP/R/ttestbayesianpairedsamples.R [hunk 13]
```diff
@@ -813,12 +829,14 @@ TTestBayesianPairedSamples <- function(dataset=NULL, options, perform="run", cal
...
-						
-						p <- try(silent= FALSE, expr= {
-						  
+					if (!is.null(errorMessage)) {
+					    
+					    plot[["error"]] <- list(error="badData", errorMessage=errorMessage)
+					    
+					} else {				
+					    
+					    p <- try(silent= FALSE, expr= {
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/ttestbayesianpairedsamples.R]: And here.

File: JASP-Engine/JASP/R/ttestbayesianpairedsamples.R [hunk 14]
```diff
@@ -896,7 +895,13 @@ TTestBayesianPairedSamples <- function(dataset=NULL, options, perform="run", cal
...
-					} else if (status$unplotable == FALSE) {
+					} 
+					
+					if (!is.null(errorMessage)){
+					    
+					    plot[["error"]] <- list(error="badData", errorMessage=errorMessage)
+					    
+					} else {
 					
 						p <- try(silent= FALSE, expr= {
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/ttestbayesianpairedsamples.R]: Here too.

**Comment by koenderks** [comment]
Tim, can you review the new changes?

File: JASP-Engine/JASP/R/ttestbayesianonesample.R [hunk 15]
```diff
@@ -605,85 +605,69 @@ TTestBayesianOneSample <- function(dataset=NULL, options, perform="run", callbac
...
-        } else {
-          
-          if (is.na(bf.raw)) {
-            
-            status[i] <- "error"
-            plottingError[i] <- "Plotting is not possible: Bayes factor could not be calculated"
-          } else if(bf.raw == Inf & (options$plotPriorAndPosterior | options$plotBayesFactorRobustness | options$plotSequentialAnalysis | options$plotSequentialAnalysisRobustness)){
-            
-            status[i] <- "error"
-            plottingError[i] <- "Plotting is not possible: Bayes factor is infinite"
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/ttestbayesianonesample.R]: Why did you delete this block?
Infinity in the BF != infinity in the dataset
Right now our messages would not be informative enough.

File: JASP-Engine/JASP/R/ttestbayesianpairedsamples.R [hunk 16]
```diff
@@ -507,24 +522,7 @@ TTestBayesianPairedSamples <- function(dataset=NULL, options, perform="run", cal
 						bf.raw <- r[["bf"]]
 						tValue[i] <- r[["tValue"]]
 						n[i] <- r[["n1"]]
-						
-						if (is.na(bf.raw)) {
```
Comment (TimKDJ) [inline JASP-Engine/JASP/R/ttestbayesianpairedsamples.R]: Same as first remark. What would the message now be if the BF is NA?

**Comment by TimKDJ** [comment]
Looks good now!
I tested it on my machine and it behaves as expected :)

There does appear to be a merge conflict. Maybe you can resolve that.
